Resume Signal Form

james-union opened the pull request:
https://github.com/flyteorg/flyteconsole/issues/587

# TL;DR
- SetSignal is still waiting to be deployed on the backend side

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
fixes https://github.com/flyteorg/flyteconsole/issues/587

olga-union commented on packages/zapp/console/src/components/Launch/LaunchForm/constants.ts:
```diff
@@ -4,13 +4,17 @@ import { BlobValue, InputType, AuthRoleTypes, AuthRoleMeta } from './types';
 export const formStrings = {
   cancel: 'Cancel',
   inputs: 'Inputs',
+  gateInput: 'Gate input',
   role: 'Role',
   submit: 'Launch',
+  resume: 'Resume',
   taskVersion: 'Task Version',
   title: 'Create New Execution',
+  resumeTitle: 'Resume Paused Execution',
   workflowVersion: 'Workflow Version',
   launchPlan: 'Launch Plan',
   interruptible: 'Interruptible',
+  viewNodeInputs: 'View node inputs',
 };
```
let's move all of the formStrings and lines 77-90 (except line 83 `workflowNoInputsString`, as it's being imported outside of the LaunchForm too) in a localization strings file for this folder

olga-union commented on packages/zapp/console/src/components/Launch/LaunchForm/ResumeForm.tsx:
```diff
@@ -0,0 +1,15 @@
+
+/** Renders the form for initiating a Launch request based on a Workflow or Task */
+export const ResumeForm: React.FC<ResumeSignalFormProps> = (props) => {
+  const [inputValueCache] = React.useState(createInputValueCache());
```
update hook import

olga-union replied:
Ok, I don't think you use `inputValueCache` anywhere in the resume form, so we don't need this wrapper - `<ResumeSignalForm />` can be used directly in its place

olga-union replied:
even better would be to rename `ResumeSignalForm` as `ResumeForm` and delete this component

olga-union commented on packages/zapp/console/src/components/Executions/contextProvider/NodeExecutionDetails/createExecutionArray.tsx:
```diff
@@ -3,7 +3,7 @@ import { getTaskDisplayType } from 'components/Executions/utils';
 import { Workflow } from 'models/Workflow/types';
 import { Identifier } from 'models/Common/types';
-import { CompiledTask } from 'models/Task/types';
+import { CompiledTask, TaskTemplate } from 'models/Task/types';
```
remove unused type import

olga-union reviewed (commented):
_Don't forget to include `flyteidl` update in this PR_ - ignore for now, looks like v1.1.22 doesn't have gate nodes

olga-union commented on packages/zapp/console/src/components/Launch/LaunchForm/types.ts:
```diff
@@ -5,8 +5,9 @@ import {
   NamedEntityIdentifier,
 } from 'models/Common/types';
-import { WorkflowExecutionIdentifier } from 'models/Execution/types';
+import { NodeExecutionIdentifier, WorkflowExecutionIdentifier } from 'models/Execution/types';
```
`NodeExecutionIdentifier` is unused

olga-union commented on packages/zapp/console/src/components/Launch/LaunchForm/useResumeFormState.ts:
```diff
@@ -0,0 +1,116 @@
+    [apiContext, formInputsRef],
+  );
+
+  const [state, sendEvent, service] = useMachine<
```
`sendEvent` is unused

olga-union commented on packages/zapp/console/src/components/Launch/LaunchForm/ResumeFormHeader.tsx:
```diff
@@ -0,0 +1,14 @@
+import { DialogTitle, Typography } from '@material-ui/core';
+import * as React from 'react';
+import { formStrings } from './constants';
+import { useStyles } from './styles';
+
+export const ResumeFormHeader: React.FC<{ title?: string }> = ({ title = '' }) => {
+  const styles = useStyles();
+  return (
+    <DialogTitle disableTypography={true} className={styles.header}>
+      <div className={styles.inputLabel}>{formStrings.resumeTitle}</div>
+      <Typography variant="h6">{title}</Typography>
+    </DialogTitle>
+  );
+};
```
This is 99% duplicate of `LaunchFormHeader`, we need to reuse the existing component and just pass down appropriate `formTitle` prop based on the place it's being rendered (formStrings.resumeTitle or formStrings.title)

olga-union commented on packages/zapp/console/src/components/Launch/LaunchForm/ResumeFormActions.tsx:
```diff
@@ -0,0 +1,85 @@
+import { Button, DialogActions, FormHelperText } from '@material-ui/core';
```
same here - 99% copy-paste, reuse `LaunchFormAction` and pass an extra prop with button title

olga-union commented on packages/zapp/console/src/components/Launch/LaunchForm/ResumeFormDialog.tsx:
```diff
@@ -0,0 +1,41 @@
+import { Dialog } from '@material-ui/core';
```
This one is ~90% similarity, so no need for a separate component too.
Next props would be optional: `id`, `compiledNode`, `nodeId`, so inside the `Dialog` you'll have a ternary `id ? <LaunchForm /> : {compiledNode && nodeId && <ResumeForm />}` or something like that

olga-union commented on packages/zapp/console/src/components/Launch/LaunchForm/useResumeFormState.ts:
```diff
@@ -0,0 +1,122 @@
+} from './types';
+import { getUnsupportedRequiredInputs } from './utils';
+
+export interface ResumeFormProps extends BaseLaunchFormProps {
+  compiledNode: CompiledNode;
+  initialParameters?: TaskInitialLaunchParameters;
+  nodeId: string;
+}
```
unused it this file, and has the same definition in `ResumeForm` component

olga-union commented on packages/zapp/console/src/components/Launch/LaunchForm/ResumeForm.tsx:
```diff
@@ -0,0 +1,23 @@
+import { ResumeSignalForm } from './ResumeSignalForm';
+import { BaseLaunchFormProps, TaskInitialLaunchParameters } from './types';
+
+export interface ResumeFormProps extends BaseLaunchFormProps {
+  compiledNode: CompiledNode;
+  initialParameters?: TaskInitialLaunchParameters;
+  nodeId: string;
+}
```
remove `export` - it's a local interface

olga-union commented on packages/zapp/console/src/components/Launch/LaunchForm/ResumeSignalForm.tsx:
```diff
@@ -0,0 +1,82 @@
+  const styles = useStyles();
+  const baseState = state as BaseInterpretedLaunchState;
+  const baseService = service as BaseLaunchService;
+  const [isError, setIsError] = React.useState<boolean>(false);
```
React. leftover

olga-union commented on packages/zapp/console/src/components/Launch/LaunchForm/LaunchFormDialog.tsx:
```diff
@@ -24,7 +28,7 @@ function getLaunchProps(id: ResourceIdentifier) {
 }
 
 export const LaunchFormDialog = (props: LaunchFormDialogProps): JSX.Element => {
```
```suggestion
export const LaunchFormDialog = ({ id, initialParameters, showLaunchForm, setShowLaunchForm, compiledNode, nodeId }: LaunchFormDialogProps): JSX.Element => {
```

olga-union commented on packages/zapp/console/src/components/Launch/LaunchForm/LaunchFormDialog.tsx:
```diff
@@ -24,7 +28,7 @@ function getLaunchProps(id: ResourceIdentifier) {
 
 export const LaunchFormDialog = (props: LaunchFormDialogProps): JSX.Element => {
-  const { id, initialParameters, showLaunchForm, setShowLaunchForm } = props;
+  const { id, initialParameters, showLaunchForm, setShowLaunchForm, compiledNode, nodeId } = props;
```
```suggestion
```

olga-union commented on packages/zapp/console/src/components/Executions/Tables/NodeExecutionActions.tsx:
```diff
@@ -122,7 +105,26 @@ export const NodeExecutionActions = ({ execution }: NodeExecutionActionsProps):
         </IconButton>
       </Tooltip>
-      {renderRerunAction()}
+      <Tooltip title={t('rerunTooltip')}>
+        <IconButton onClick={rerunIconOnClick}>
+          <RerunIcon />
+        </IconButton>
+      </Tooltip>
+      <LaunchFormDialog
+        id={id as ResourceIdentifier}
+        initialParameters={initialParameters}
+        showLaunchForm={showLaunchForm}
+        setShowLaunchForm={setShowLaunchForm}
+      />
```
I think this should be wrapped in id and initialParameters presence check
```suggestion
      {id && initialParameters && (<Tooltip title={t('rerunTooltip')}>
        <IconButton onClick={rerunIconOnClick}>
          <RerunIcon />
        </IconButton>
      </Tooltip>
      <LaunchFormDialog
        id={id as ResourceIdentifier}
        initialParameters={initialParameters}
        showLaunchForm={showLaunchForm}
        setShowLaunchForm={setShowLaunchForm}
      />
      )}
```

olga-union commented on packages/zapp/console/src/components/Launch/LaunchForm/LaunchInterruptibleInput.tsx:
```diff
@@ -74,14 +74,14 @@ export const LaunchInterruptibleInputImpl: React.ForwardRefRenderFunction<
   const getInterruptibleLabel = () => {
```
Can you please add TODO above this function to cover all text variants in localization conditional?

olga-union commented on packages/zapp/console/src/components/Launch/LaunchForm/ResumeForm.tsx:
```diff
@@ -0,0 +1,23 @@
+  nodeId: string;
+}
+
+/** Renders the form for initiating a Launch request based on a Workflow or Task */
```
This description doesn't match the purpose of the component

olga-union commented on packages/zapp/console/src/components/Launch/LaunchForm/ResumeSignalForm.tsx:
```diff
@@ -0,0 +1,85 @@
+  nodeId: string;
+}
+
+/** Renders the form for initiating a Launch request based on a Task */
```
this one is related to original LaunchForm too 😄

olga-union commented on packages/zapp/console/src/components/Launch/LaunchForm/ResumeSignalForm.tsx:
```diff
@@ -0,0 +1,85 @@
+
+  // Any time the inputs change (even if it's just re-ordering), we must
+  // change the form key so that the inputs component will re-mount.
+  const formKey = React.useMemo<string>(() => {
```
update import

olga-union reviewed (commented):
Nice job! Few tiny things left, and don't forget to checkout `package.json`s you accidentally committed 😃

james-union commented:
@olga-union I've fixed your comments. Just FYI, the flyteidl version is not yet updated because if it points to 1.1.14, it fails to build as `Core.SignalIdentifier` is not defined. 
For the tests, I've fixed `PausedTasksComponent` & `NodeExecutionActions`

Still working on `ResumeSignalForm`
